### ISSUE TITLE
batcheval: 24.3 microbench regression in `BenchmarkRefreshRange`

### ISSUE THREAD

**Issue by tbg** [cockroachdb/cockroach#134420]
[profs.zip](https://github.com/user-attachments/files/17649628/profs.zip)

```
./dev bench --ignore-cache --stream-output --bench-mem ./pkg/kv/kvserver/batcheval/ --filter 'RefreshRange/mixed-case/refresh_window=.0.00,0.00.' --test-args '-test.memprofile=mem.pb.gz -test.cpuprofile=cpu.pb.gz -test.benchtime=50000x -test.cpu=8'
```

Clearly the new one indeed is much slower (I think the spreadsheet says ~75%, and I verified that), but it's unclear why. The additional time seems to be spent in `iterator.SeekGE`.

old

![image](https://github.com/user-attachments/assets/100b756a-7045-4b59-aeaf-9daae1a07725)

new

![image](https://github.com/user-attachments/assets/2c987e7b-ddd1-4930-b5ac-42a5fa25d88c)

There's something weird about the `refresh_window=[N,N]` case. Note how `[0,0]` is way slower than `[0,anything-else]`. Similar, how `[50,50]` is by far the slowest of the `[50,*]` family, etc.

```
$ ./dev bench --ignore-cache --stream-output --bench-mem ./pkg/kv/kvserver/batcheval/ --filter 'RefreshRange/mixed-case/' --test-args '-test.memprofile=mem.pb.gz -test.cpuprofile=cpu.pb.gz -test.benchtime=3s -test.cpu=8' 2>&1 | grep --line-buffered -E '^Benchmark' | tee bench.txt
BenchmarkRefreshRange/mixed-case/refresh_window=[0.00,0.00]-8              16135            222172 ns/op            1325 B/op         20 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[0.00,50.00]-8            209877             16824 ns/op            1398 B/op         23 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[0.00,75.00]-8            208724             16817 ns/op            1399 B/op         23 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[0.00,95.00]-8            170229             20697 ns/op            1398 B/op         23 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[0.00,99.00]-8            180392             19752 ns/op            1398 B/op         23 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[50.00,50.00]-8               25         132782851 ns/op            7538 B/op         38 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[50.00,75.00]-8           200251             17568 ns/op            1404 B/op         24 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[50.00,95.00]-8           168374             21018 ns/op            1404 B/op         24 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[50.00,99.00]-8           177073             20014 ns/op            1404 B/op         24 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[75.00,75.00]-8               26         128251366 ns/op            4628 B/op         38 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[75.00,95.00]-8           168642             20962 ns/op            1404 B/op         24 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[75.00,99.00]-8           177039             20013 ns/op            1404 B/op         24 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[95.00,95.00]-8           113584             31360 ns/op            1321 B/op         20 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[95.00,99.00]-8           211741             16561 ns/op            1404 B/op         24 allocs/op
BenchmarkRefreshRange/mixed-case/refresh_window=[99.00,99.00]-8             1128           3128315 ns/op            1359 B/op         20 allocs/op
```

The mixed-case benchmark family roughly writes 1 million (separate) keys in random order, key `i` at timestamp `5ns*(1+i)`. The engine is set up with an `LBaseMaxBytes` of `256` to create a more interesting LSM. Then, it benchmarks `RefreshRange` with a timestamp range that covers the given percentage of cases.

For example, `[0,0]` translates to refreshing from `ts=5ns` to `ts=5ns+1 logical`. It makes zero sense to me that that would take 222us per op, whereas refreshing over essentially all of the data would take around 20us.

But here's the kicker, the above table is from b20c617197fc13f08ae9be59ec40254c17c9e0f5, i.e. the "old" commit in the microbench regression. There's little point in musing about why it's gotten almost twice as bad (and it has - ~388 micros for the `[0,0]` case) if we don't understand either result in the first place.

Jira issue: CRDB-44073

**Comment by tbg** [comment]
I bisected this down to e89aab636ca666e4c1f788b4739d2dfa03255c1d using

```
rm -rf ~/.cache/crdb-test-fixtures; ./dev bench --ignore-cache --stream-output --bench-mem ./pkg/kv/kvserver/batcheval/ --filter 'RefreshRange/mixed-case/refresh_window=.0.00,0.00.' --test-args '-test.memprofile=mem.pb.gz -test.cpuprofile=cpu.pb.gz -test.benchtime=50000x -test.cpu=8'
```

Since there weren't too many steps, I just eyeballed - "good" on my gceworker was <<300us, bad was >> 300us. The other `[N,N]` cases were also penalized. Note that these are the cases in which `RefreshRange` returns success, i.e. the happy ones in which a txn can continue to run. In the other cases, it finds a key that allows it to return early with an error (and the txn has to re-do its reads).

Note that the test builds up an LSM (see setupData[^1]) so perhaps the shape of the LSM has changed; I haven't checked.

[^1]: https://github.com/cockroachdb/cockroach/blob/e8f6410ecd99591fd6103dd8461ea3fb0361f8d6/pkg/kv/kvserver/batcheval/cmd_refresh_range_bench_test.go#L209

go.mod: bump Pebble to c6b097e7e6a0

Changes:

 * [`c6b097e7`](https://github.com/cockroachdb/pebble/commit/c6b097e7) sstable: add jemalloc size classes to tests
 * [`b07f0b23`](https://github.com/cockroachdb/pebble/commit/b07f0b23) db: fix Options.Parse handling of unrecognized comparer
 * [`9f164613`](https://github.com/cockroachdb/pebble/commit/9f164613) db: fix a few instances of the wrong table format
 * [`dd26bf0e`](https://github.com/cockroachdb/pebble/commit/dd26bf0e) db: delete compactions output in case of late cancellation
 * [`1ed55247`](https://github.com/cockroachdb/pebble/commit/1ed55247) block: rewrite block flush logic
 * [`887598ae`](https://github.com/cockroachdb/pebble/commit/887598ae) sstable: set w.indexBlockSize in colBlkWriter
 * [`030afd26`](https://github.com/cockroachdb/pebble/commit/030afd26) colblk: use UnsafeOffsets.At2 in another instance
 * [`b3993ff6`](https://github.com/cockroachdb/pebble/commit/b3993ff6) db: persist Experimental.EnableColumnarBlocks in OPTIONS file
 * [`c719c4e5`](https://github.com/cockroachdb/pebble/commit/c719c4e5) db: support multiple KeySchemas

**Comment by jbowens** [comment]
Thanks for the bisect @tbg!

Diffing CPU profiles before and after:

```
Showing nodes accounting for 0.60s, 1.18% of 50.64s total
Showing top 30 nodes out of 883
      flat  flat%   sum%        cum   cum%
    -2.82s  5.58%  5.58%     -2.82s  5.58%  runtime/internal/syscall.Syscall6
     1.10s  2.17%  3.41%      1.10s  2.17%  encoding/binary.Uvarint (inline)
     1.02s  2.02%  1.39%      1.21s  2.39%  github.com/cockroachdb/pebble/sstable/rowblk.(*Iter).readEntry
    -0.79s  1.55%  2.95%     -0.60s  1.18%  runtime.cgocall
     0.69s  1.36%  1.59%      1.46s  2.89%  github.com/cockroachdb/pebble/sstable.(*singleLevelIterator[go.shape.struct { github.com/cockroachdb/pebble/sstable/rowblk.iter github.com/cockroachdb/pebble/sstable/rowblk.Iter },go.shape.*github.com/cockroachdb/pebble/sstable/rowblk.IndexIter,go.shape.eb7c958227f421b2b34e33f9383baa8eb4cd132ad8ad1af0ac397c32eec3846d,go.shape.*github.com/cockroachdb/pebble/sstable/rowblk.Iter]).loadDataBlock
     0.68s  1.34%  0.24%      1.47s  2.90%  github.com/cockroachdb/pebble/sstable/rowblk.(*IndexIter).BlockHandleWithProperties
     0.41s  0.82%  0.57%      0.42s  0.83%  github.com/golang/snappy.decode
    -0.40s   0.8%  0.23%     -0.36s   0.7%  github.com/cockroachdb/pebble/internal/batchskl.(*Skiplist).findSpliceForLevel
     0.34s  0.68%  0.45%      1.60s  3.16%  github.com/cockroachdb/pebble/sstable/rowblk.(*Iter).Next
    -0.30s   0.6%  0.14%     -0.30s   0.6%  hash/crc32.castagnoliSSE42Triple
     0.29s  0.58%  0.43%      1.30s  2.57%  github.com/cockroachdb/pebble/sstable.(*BlockPropertiesFilterer).intersectsFilter
     0.29s  0.58%  1.01%      0.88s  1.74%  github.com/cockroachdb/pebble/sstable.decodeBlockInterval
    -0.24s  0.47%  0.54%     -0.23s  0.46%  github.com/golang/snappy.encodeBlock
     0.24s  0.47%  1.01%      1.77s  3.50%  github.com/cockroachdb/pebble/sstable.(*BlockPropertiesFilterer).intersects
    -0.21s  0.41%   0.6%     -0.78s  1.55%  github.com/cockroachdb/pebble/internal/manual.New
     0.20s   0.4%     1%      0.60s  1.19%  github.com/cockroachdb/pebble/sstable/block.DecodeHandle
    -0.20s  0.39%  0.61%     -5.07s 10.01%  github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval_test.setupData.func1.2
     0.18s  0.35%  0.96%      0.20s   0.4%  github.com/cockroachdb/pebble/sstable/rowblk.(*Iter).Invalidate
     0.17s  0.35%  1.30%      0.78s  1.54%  github.com/cockroachdb/pebble/sstable/block.DecodeHandleWithProperties
     0.17s  0.34%  1.64%      0.27s  0.53%  github.com/cockroachdb/pebble/sstable.(*blockPropertiesDecoder).Next
    -0.15s   0.3%  1.33%     -0.08s  0.16%  math/rand.(*Rand).Int31n
    -0.15s   0.3%  1.03%     -0.94s  1.85%  github.com/cockroachdb/pebble/internal/cache.newValue
    -0.14s  0.29%  0.75%     -0.22s  0.43%  github.com/cockroachdb/pebble/internal/arenaskl.(*Skiplist).findSpliceForLevel
     0.14s  0.28%  1.02%         1s  1.97%  github.com/cockroachdb/pebble/sstable.(*BlockIntervalFilter).Intersects
     0.14s  0.27%  1.30%      0.14s  0.27%  runtime.duffcopy
     0.14s  0.27%  1.57%      6.12s 12.08%  github.com/cockroachdb/pebble/sstable.(*singleLevelIterator[go.shape.struct { github.com/cockroachdb/pebble/sstable/rowblk.iter github.com/cockroachdb/pebble/sstable/rowblk.Iter },go.shape.*github.com/cockroachdb/pebble/sstable/rowblk.IndexIter,go.shape.eb7c958227f421b2b34e33f9383baa8eb4cd132ad8ad1af0ac397c32eec3846d,go.shape.*github.com/cockroachdb/pebble/sstable/rowblk.Iter]).skipForward
    -0.11s  0.21%  1.36%     -0.19s  0.38%  github.com/cockroachdb/cockroach/pkg/util/randutil.RandBytes
     0.10s  0.19%  1.55%      0.10s  0.19%  runtime.nanotime (partial-inline)
    -0.10s  0.19%  1.36%     -0.20s  0.39%  github.com/cockroachdb/pebble/sstable/rowblk.(*Iter).SeekGE
    -0.09s  0.19%  1.18%      4.53s  8.95%  github.com/cockroachdb/pebble.(*Iterator).findNextEntry
```

I'm suspicious that the change to the block-splitting + size class heuristics resulted in more data blocks, leading to more index block entries and more decoding of block properties in order to skip over the same amount of data. The columnar blocks implementation uses the same logic now, but iteration over its index blocks should be faster which may contribute to 'papering over' the regression. I'm going to look into the generated fixtures to see if the data block and index block size distributions vary.

**Comment by jbowens** [comment]
Yeah, @RaduBerinde :

In a run on e89aab636ca666e4c1f788b4739d2dfa03255c1d^, the generated fixture has four sstables with 2543, 156, 129 and 14 data blocks.

In a run on e89aab636ca666e4c1f788b4739d2dfa03255c1d, the generated fixture has four sstables with 4069, 250, 206 and 23 data blocks.

I'm inclined to think we should accept the regression here with the understanding that in practice, we're improving our block cache hit ratio for the same amount of memory by reducing memory fragmentation and improving block cache hits is likely more important to overall performance than reducing the amount of varint decoding performed when skipping over whole data blocks.

**Comment by jbowens** [comment]
Hrm, these blocks do seem too small with uncompressed sizes around 20kb. Maybe we're misconfiguring the size classes?

**Comment by jbowens** [comment]
@RaduBerinde ; am I reading correctly that there was a behavior change in cockroachdb/pebble#4060 in which we began to consider flushing as soon as we hit the SizeClassAwareThreshold-adjusted size, even if the next key value would _not_ put us over the BlockSize? I think this change caused us to begin to flush as soon as we hit the lowest size class threshold within the SizeClassAwareThreshold, which is around 20kb.

**Comment by jbowens** [comment]
Removing the release-blocker label and adding GA-blocker. We effectively reduced the block size to 60% of its previous size. We're going to re-evaluate this logic before .0.

**Comment by jbowens** [comment]
24.3 was fixed by #134696.
